Scripts/Data/Collection/ObservableCounter.cs

```diff
@@ -0,0 +1,114 @@
+{
+    using UnityEngine;
+    using UnityEngine.Events;
+    using System.Linq;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Allows observing counting the amount of attempts an element is added or removed from a <see cref="Dictionary{TKey, TValue}"/>.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the elements to count.</typeparam>
+    /// <typeparam name="TEvent">The <see cref="UnityEvent"/> type to use.</typeparam>
+    public class ObservableCounter<TElement, TEvent> : MonoBehaviour where TEvent : UnityEvent<TElement>, new()
+    {
+        /// <summary>
+        /// Emitted when a searched element is found.
+        /// </summary>
+        public TEvent ElementAdded = new TEvent();
+        /// <summary>
+        /// Emitted when a searched element is not found.
+        /// </summary>
+        public TEvent ElementRemoved = new TEvent();
+
+        /// <summary>
+        /// The elements being counted.
+        /// </summary>
+        public Dictionary<TElement, int> ElementsCounter
+        {
+            get;
+            protected set;
+        } = new Dictionary<TElement, int>();
+
+        /// <summary>
+        /// Increases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void IncreaseCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null)
+            {
+                return;
+            }
+
+            if (!ElementsCounter.ContainsKey(element))
+            {
+                ElementsCounter.Add(element, 0);
+            }
+
+            ElementsCounter[element]++;
+            if (ElementsCounter[element] == 1)
+            {
+                ElementAdded?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Decreases the count on the given element.
+        /// </summary>
+        /// <param name="element">The element to count.</param>
+        public virtual void DecreaseCount(TElement element)
+        {
+            int currentValue = 0;
+            if (!isActiveAndEnabled || element == null || !ElementsCounter.TryGetValue(element, out currentValue) || currentValue <= 0)
+            {
+                return;
+            }
+
+            ElementsCounter[element]--;
+            if (ElementsCounter[element] <= 0)
+            {
+                ElementsCounter.Remove(element);
+                ElementRemoved?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Removes the element from the counter.
+        /// </summary>
+        /// <param name="element">The element to clear.</param>
+        public virtual void RemoveFromCount(TElement element)
+        {
+            if (!isActiveAndEnabled || element == null || !ElementsCounter.ContainsKey(element))
+            {
+                return;
+            }
+
+            ElementsCounter.Remove(element);
```

Comment:
```suggestion
            if (ElementsCounter.Remove(element))
            {
```

Reply:
Could this method not just be simplified to:

```
public virtual void RemoveFromCount(TElement element)
{
    if (!isActiveAndEnabled || element == null || !ElementsCounter.Remove(element))
    {
        return;
    }
    ElementRemoved?.Invoke(element);
}
```

if it's not active or enabled or the element is null or when i try and remove the element it doesn't find one to remove then bomb out early, otherwise invoke the method.

Reply:
👍